src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs

```diff
@@ -1638,5 +1638,65 @@ class Source2
                 );
             var comp = CompileAndVerify(compilation, expectedOutput: expectedOutput);
         }
+
+        [Fact, WorkItem(40533, "https://github.com/dotnet/roslyn/issues/40533")]
+        public void DisallowDesignatorsUnderNotAndOr()
+        {
+            var source = @"
+class C
+{
+    void Good(object o)
+    {
+        if (o is int and 1) { }
+        if (o is int x1 and 1) { }
+        if (o is int x2 and (1 or 2)) { }
+        if (o is 1 and int x3) { }
+        if (o is (1 or 2) and int x4) { }
+        if (o is not (1 or 2) and int x5) { }
+    }
+
+    void Bad(object o)
+    {
+        if (o is int y1 or 1) { }
+        if (o is int y2 or (1 or 2)) { }
+        if (o is 1 or int y3) { }
+        if (o is (1 or 2) or int y4) { }
+        if (o is not int y5) { }
+        if (o is not (1 and int y6)) { }
```

Comment:
Consider other nesting cases: `Point { X: var x } or Animal` and `Point (var x, _) or Animal`